src/utils/__tests__/async-test.js

```diff
@@ -1,14 +1,47 @@
 /* @flow strict-local */
 import { sleep, tryUntilSuccessful } from '../async';
+import { Lolex } from '../../__tests__/aux/lolex';
 
 describe('sleep', () => {
+  let lolex: Lolex;
+
+  beforeAll(() => {
+    lolex = new Lolex();
+  });
+
+  afterAll(() => {
+    lolex.dispose();
+  });
+
+  afterEach(() => {
+    // clear any unset timers
+    lolex.clearAllTimers();
+  });
+
   test('waits for a given time in milliseconds', async () => {
     const expectedMs = 1000;
-    const start = Date.now();
-    await sleep(expectedMs);
-    const durationMs = Date.now() - start;
-    expect(expectedMs).toBeLessThanOrEqual(durationMs);
-    expect(durationMs).toBeLessThan(10 * expectedMs);
+
+    const expectationPromise = async () => {
+      const start = Date.now();
+      await sleep(expectedMs);
+      const durationMs = Date.now() - start;
+
+      expect(expectedMs).toBeLessThanOrEqual(durationMs);
+      // cannot be guaranteed: arbitrary delays are possible
+      // expect(durationMs).toBeLessThan(10 * expectedMs);
+
+      return true;
+    };
```

Comment:
I'd really love to avoid nesting stuff inside this promise, to the extent we can -- it makes the control flow a lot more complicated to reason about. Or if it really does need to be inside here to get us things we want, I'd like to understand clearly why that's the case, which I don't yet.

For a first step: what happens if we pull the `expect` call outside the promise, to be done after `await`ing it? And have this function return `durationMs`. Does that change the test's behavior, and if so how?

(Maybe this interacts with the answer to my question about "if this stalls".)

Reply:
No, that works just fine. Shuffled appropriately.